src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2180,18 +2188,30 @@ void ShenandoahHeap::op_degenerated(ShenandoahDegenPoint point) {
       //
       // Note that we can only do this for "outside-cycle" degens, otherwise we would risk
       // changing the cycle parameters mid-cycle during concurrent -> degenerated handover.
+
+      // Degenerated from concurrent mark roots, reset for STW mark
+      if (is_concurrent_mark_in_progress()) {
+        ShenandoahConcurrentMark::cancel();
+        set_concurrent_mark_in_progress(false);
+      }
+
       set_unload_classes(heuristics()->can_unload_classes());
 
       op_reset();
 
-      op_init_mark();
-      if (cancelled_gc()) {
-        op_degenerated_fail();
-        return;
+      // STW root scan
+      {
+        assert(!has_forwarded_objects(), "Should not have forwarded heap");
+        ShenandoahSTWMark mark(false /*full_gc*/);
+        mark.mark();
+        assert(!cancelled_gc(), "STW mark can not OOM");
       }
-
     case _degenerated_mark:
-      op_final_mark();
+      if (point == _degenerated_mark) {
+        finish_mark();
+      }
```

Comment:
So if we don't call `finish_mark`, do we ever call  `set_concurrent_mark_in_progress(false);` and `mark_complete_marking_context();`?

Reply:
This was not answered, and I don't see relevant follow-up changes. Is this a bug?

Reply:
No, I don't think so.

For fall through case, STWMark sets both flags. 

For degen case, if we pass _degnerated_mark point, op_final_mark calls finish_mark to set them.